# NUMPY - Multidimensional Data Arrays

It is a package that provide high-performance vector, matrix and higher-dimensional data structures for Python. NumPy brings the computational power of languages like C and Fortran to Python, a language much easier to learn and use.

### Import Numpy Library 

In [1]:
import numpy as np

In [2]:
np.__version__

'1.19.5'

### Getting Help

In [3]:
np.info(np.array)

array(object, dtype=None, *, copy=True, order='K', subok=False, ndmin=0)

Create an array.

Parameters
----------
object : array_like
    An array, any object exposing the array interface, an object whose
    __array__ method returns an array, or any (nested) sequence.
dtype : data-type, optional
    The desired data-type for the array.  If not given, then the type will
    be determined as the minimum type required to hold the objects in the
    sequence.
copy : bool, optional
    If true (default), then the object is copied.  Otherwise, a copy will
    only be made if __array__ returns a copy, if obj is a nested sequence,
    or if a copy is needed to satisfy any of the other requirements
    (`dtype`, `order`, etc.).
order : {'K', 'A', 'C', 'F'}, optional
    Specify the memory layout of the array. If object is not an array, the
    newly created array will be in C order (row major) unless 'F' is
    specified, in which case it will be in Fortran order (column major).
    If object 

In [ ]:
np.array()

## What is Array?

![](image/array.png)

### List vs Numpy 

In [4]:
# Python List
list1 = [1, 2, 3, 4, 5]
list2 = [i*2 for i in range(1, 6)] # 1 2 3 4 5 

print('list1:', list1)
print('list2:', list2)

list1: [1, 2, 3, 4, 5]
list2: [2, 4, 6, 8, 10]


In [5]:
print('tipedata dari list1:', type(list1))

tipedata dari list1: <class 'list'>


In [6]:
list1 + list2

[1, 2, 3, 4, 5, 2, 4, 6, 8, 10]

In [7]:
list3 = []
for a, b in zip(list1, list2):
    list3.append(a+b)
list3

[3, 6, 9, 12, 15]

In [8]:
list1 * list2

TypeError: can't multiply sequence by non-int of type 'list'

In [9]:
list3 = []
for a, b in zip(list1, list2):
    list3.append(a*b)
list3

[2, 8, 18, 32, 50]

In the `numpy` package the terminology used for vectors, matrices and higher-dimensional data sets is *array*. 



In [10]:
# numpy
np1 = np.array([1, 2, 3, 4, 5])
np2 = np.array([i*2 for i in range(1, 6)])

print(np1)
print(np2)

[1 2 3 4 5]
[ 2  4  6  8 10]


In [11]:
type(np1)

numpy.ndarray

In [12]:
np1 + np2

array([ 3,  6,  9, 12, 15])

In [13]:
np1 * np2

array([ 2,  8, 18, 32, 50])

In [14]:
np1 ** np2

array([      1,      16,     729,   65536, 9765625], dtype=int32)

In [16]:
list1 = list(range(1, 1001))
', '.join([str(i) for i in list1])

'1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 22

In [17]:
len(list1)

1000

In [18]:
import sys

In [19]:
a = 1 

In [20]:
type(a)

int

In [21]:
sys.getsizeof(a)

28

In [22]:
sys.getsizeof(a) * len(list1)

28000

In [23]:
np1 = np.arange(1, 1001)

In [26]:
len(np1)

1000

In [25]:
np1.size

1000

In [27]:
np1.itemsize

4

In [28]:
np1.size * np1.itemsize

4000

In [29]:
size = 10_000_000

In [33]:
list1 = list(range(size))
list2 = list(range(size))
np1 = np.arange(size)
np2 = np.arange(size)

In [30]:
import time 

In [34]:
mulai = time.time()
result = [a+b for a, b in zip(list1, list2)]
print('python list membutuhkan waktu:', time.time() - mulai, 's')

python list membutuhkan waktu: 1.4090018272399902 s


In [35]:
mulai = time.time()
result = np1 + np2
print('python numpy membutuhkan waktu:', time.time() - mulai, 's')

python numpy membutuhkan waktu: 0.17798948287963867 s


## Creating `numpy` arrays

There are some ways to initialize new numpy arrays:
* a Python list or tuples
* using functions that are dedicated to generating numpy arrays, such as `arange`, `linspace`, etc.
* reading data from files

### Lists

In [36]:
# vector: the argument to the array function is a list
v = np.array([1, 2, 3, 4, 5])

v

array([1, 2, 3, 4, 5])

In [37]:
type(v)

numpy.ndarray

In [38]:
# matrix: the argument to the array function is a nested list
m = np.array([[1, 2, 3], [4, 5, 6]])

m

array([[1, 2, 3],
       [4, 5, 6]])

The `v` and `m` objects are both of the type `ndarray` that the `numpy` module provides.

In [39]:
type(v), type(m)

(numpy.ndarray, numpy.ndarray)

In [40]:
a = [1,2,3]

In [41]:
type(a)

list

The difference between the `v` and `m` arrays is only their shapes. We can get information about the shape of an array by using the `ndarray.shape` property.

In [42]:
v.shape

(5,)

In [43]:
m.shape

(2, 3)

The number of elements in the array is available through the `ndarray.size` property

In [44]:
m.size

6

`numpy.ndarray` looks very similiar to the `list`. So, why not use the list instead?
`numpay.ndarray` is used for several reason:
1. Lists are very general. They can contain any kind of object. They do not support mathematical functions such as matrix and dot multiplication, etc. 
2. Numpy arrays are statically typed and homogenous. The type of the elements is determined when the array is created
3. Numpy arrays are memory efficient
4. It is fast for implementation of mathematical function

We can see the type of data of an array using `dtype`

In [45]:
m.dtype

dtype('int32')

If we want, we can explicitly define the type of the array data when we create it, using the `dtype` keyword argument: 

In [47]:
m = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float)

m

array([[1., 2., 3.],
       [4., 5., 6.]])

Common data types that can be used with `dtype` are: `int`, `float`, `complex`, `bool`, `object`, etc.

### Create Matrix Zeros

In [48]:
# One dimension

zeros_matrix = np.zeros(5)

zeros_matrix

array([0., 0., 0., 0., 0.])

In [49]:
#two dimension

zeros_matrix2 = np.zeros(5,2)

TypeError: Cannot interpret '2' as a data type

In [50]:
# should  be in tuple format

zeros_matrix2 = np.zeros((5,2)) # 5 rows, 2 columns
zeros_matrix2 

array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

### Matrix ones

In [51]:
#one dimension

matrix_ones = np.ones(5)
matrix_ones 

array([1., 1., 1., 1., 1.])

In [54]:
m2 = np.ones((5, 5), dtype=int)
m2 

array([[1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1]])

In [53]:
#3 dimension

matrix_ones2 = np.ones((3, 4, 2)) #3 rows, 4 columns, 2 depth
matrix_ones2

array([[[1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.]],

       [[1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.]],

       [[1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.]]])

---

## > Exercise 1

1. Create a matrix from a list which has 4 rows and 3 columns

In [55]:
n1 = np.array([[1, 2, 3], 
               [4, 5, 6],
               [7, 8, 9],
               [10, 11, 12]])
n1

array([[ 1,  2,  3],
       [ 4,  5,  6],
       [ 7,  8,  9],
       [10, 11, 12]])

In [56]:
n1.shape

(4, 3)

2. Create the following matrix
![](image/lat11.png)

In [78]:
n2 = np.array([[2, 7, 12, 0],
               [3, 9, 3, 4],
               [4, 0, 1, 3]])

brs, klm = n2.shape
for i in range(0, brs):
    for j in range(0, klm):
        print(f'{n2[i, j]:2} ', end='')
    print()


 2  7 12  0 
 3  9  3  4 
 4  0  1  3 


In [58]:
n2.shape

(3, 4)

In [60]:
n2.size

12

3. Create a 2D matrix with size of 10

In [63]:
n3 = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

In [65]:
n3.size

10

In [66]:
n4 = np.array([[1, 2, 3, 4, 5], [6, 7, 8, 9, 10]])

In [67]:
n4.size

10

In [70]:
n5 = np.zeros(10)
n5.size

10

In [71]:
n6 = np.ones((2, 5))
n6.size

10

4. Create a 3D matrix of ones which has 2 rows, 3 columns, and 3 depth

In [72]:
n7 = np.zeros((2, 3, 3))
n7

array([[[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]],

       [[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]]])

In [73]:
n7.shape

(2, 3, 3)

In [74]:
n8 = np.array([[[1, 2, 3],
                [4, 5, 6],
                [7, 8, 9]],
               [[1, 2, 3],
                [4, 5, 6],
                [7, 8, 9]]])
n8

array([[[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]],

       [[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]]])

In [75]:
n8.shape

(2, 3, 3)

5. Make the following arrays from zeros arrays and with for loops
![](image/exercise1.png)

In [79]:
n9 = np.zeros((5, 3))
n9

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [80]:
for line in n9:
    print(line + 2)

[2. 2. 2.]
[2. 2. 2.]
[2. 2. 2.]
[2. 2. 2.]
[2. 2. 2.]


---

### Using array-generating functions

For larger arrays it is inpractical to initialize the data manually, using explicit python lists. Instead we can use one of the many functions in `numpy` that generate arrays of different forms. Some of the more common are:

**arange**

In [81]:
# create a range

x = np.arange(10)

x

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [82]:
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9


In [83]:
# create a range

x = np.arange(10, 20) # arguments: start, stop

x

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19])

In [84]:
# create a range

x = np.arange(10, 20, 2) # arguments: start, stop, step

x

array([10, 12, 14, 16, 18])

In [85]:
x = np.arange(-1, 1, 0.1)

x

array([-1.00000000e+00, -9.00000000e-01, -8.00000000e-01, -7.00000000e-01,
       -6.00000000e-01, -5.00000000e-01, -4.00000000e-01, -3.00000000e-01,
       -2.00000000e-01, -1.00000000e-01, -2.22044605e-16,  1.00000000e-01,
        2.00000000e-01,  3.00000000e-01,  4.00000000e-01,  5.00000000e-01,
        6.00000000e-01,  7.00000000e-01,  8.00000000e-01,  9.00000000e-01])

The number 9.00000000e-01 already is a floating point number.
It's written in scientific notation and is equivalent to 9 * 10**-1 or 0.9.

#### linspace

In [86]:
# using linspace, both end points ARE included
np.linspace(0, 10, 10) #Unlike arange that uses step, linspace uses the number of sample

array([ 0.        ,  1.11111111,  2.22222222,  3.33333333,  4.44444444,
        5.55555556,  6.66666667,  7.77777778,  8.88888889, 10.        ])

In [88]:
np.linspace(1, 10, 4)

array([ 1.,  4.,  7., 10.])

#### random data

In [89]:
from numpy import random

In [90]:
#uniform random numbers in [0,1]
random.rand(5,5)

array([[0.39283043, 0.67957403, 0.69245126, 0.86140359, 0.30487672],
       [0.39015199, 0.92581658, 0.07892693, 0.47111119, 0.99827777],
       [0.34249275, 0.5286118 , 0.16353668, 0.44103441, 0.70567906],
       [0.32104089, 0.74164822, 0.3324951 , 0.75030791, 0.32979117],
       [0.00335814, 0.83312407, 0.54756389, 0.84120659, 0.41716525]])

In [91]:
# standard normal distributed random numbers
x = random.randn(5,5)
x

array([[ 0.0808118 ,  0.0800927 , -1.0548204 , -0.73866915, -0.16178725],
       [-1.40392132,  1.50627446, -0.39736345,  0.27483282,  0.01502622],
       [ 1.03126595, -0.67503411,  1.21935214, -0.34333602, -1.05221376],
       [-0.77300355, -0.92878116,  0.21648715,  1.49064323,  0.11411895],
       [ 0.21080725,  0.20445509, -0.30557003,  0.7082922 , -1.53403507]])

In [92]:
x.dtype

dtype('float64')

In [93]:
x = np.ones(2, dtype = np.int64)
x

array([1, 1], dtype=int64)

In [94]:
random.randint(10)

2

In [95]:
random.randint(2, 10, size=4)

array([5, 9, 4, 8])

In [96]:
random.randint(2, 10, size=(4,2,2))

array([[[4, 7],
        [5, 6]],

       [[6, 2],
        [5, 2]],

       [[5, 7],
        [3, 2]],

       [[9, 9],
        [2, 2]]])

---

## Exercise 2

1. Generate a 1-D array containing 5 random integers from 0 to 100:

2. Generate a 2-D array with 3 rows, each row contains 5 random integers from 0 to 100

3. Generate a 1-D array of 30 evenly spaced elements between 1.5 and 5.5, inclusive.